layers: Fix references in push constants

ncesario-lunarg opened the pull request:
This is a first pass at handling push constants using the
PhysicalStorageBuffer storage class.

This is for resolving Github issues #2465 and #2467.

ncesario-lunarg commented on scripts/spirv_validation_generator.py:
```diff
@@ -337,6 +337,8 @@ def extensionStruct(self):
                 if name in self.extensionExcludeList:
                     output += '    // Not found in current SPIR-V Headers\n    //'
                 output += '    {\"' + name + '\", ' + self.createMapValue(name, enable, True) + '},\n'
+        output += '\n    // Extensions not automatically generated\n'
+        output += '    {"GL_EXT_scalar_block_layout", {0, nullptr, &DeviceExtensions::vk_ext_scalar_block_layout, ""}},\n'
```
@sfricke-samsung not sure if I'm missing something here, but both tests fail if I don't add this as they both use this extension. I don't see this extension anywhere in the spirv xml (I did not check the json, though).

sfricke-samsung replied:
So what error you getting? 

`GL_EXT_scalar_block_layout` is not a SPV extension therefor it won't be found/generated from the XML. I guess there is no explict VU anywhere saying you can't using `GL_EXT_scalar_block_layout`  with `vk_ext_scalar_block_layout` not enabled (maybe it should though 🤔 )

ncesario-lunarg replied:
I get the following for the `VkPositiveLayerTest.PhysicalStorageBufferAddressesCustomCompiler` test, which does enable [VK_EXT_scalar_block_layout](https://github.com/KhronosGroup/Vulkan-ValidationLayers/pull/2497/files#diff-396c1f357ae46b5a18451f12fc652898336d51f77a03938eea969f6d0783a173R10908):
```
Expected to succeed but got error: Validation Error: [ VUID-VkShaderModuleCreateInfo-pCode-04146 ] Object 0: handle = 0x226b5a41060, type = VK_OBJECT_TYPE_DEVICE; | MessageID = 0xd80a42ae | vkCreateShaderModule(): A SPIR-V Extension (GL_EXT_scalar_block_layout) was declared that is not supported by Vulkan. The Vulkan spec states: pCode must not declare any SPIR-V extension that is not supported by the API, as described by the Extension section of the SPIR-V Environment appendix (https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/vkspec.html#VUID-VkShaderModuleCreateInfo-pCode-04146)
```

Oddly, though, the `VkPositiveLayerTest.PhysicalStorageBuffer` test does **not** fail when removing the spirv_validation_helper "hack." Not sure why that is, though.

sfricke-samsung replied:
... so I can't tell, but curious if the current layers will throw an error for all `GL_` extensions in the SPV

```
OpExtension "GL_EXT_scalar_block_layout"
```

If that is the case, well that isn't good 😟 

The spec says

> VUID-VkShaderModuleCreateInfo-pCode-04146
> 
> pCode must not declare any SPIR-V extension that is not supported by the API, as described by the Extension section of the SPIR-V Environment appendix

So if this is the case, I think the correct thing would be check the `OpExtension` value for `SPV_` before checking against the generated list

ncesario-lunarg replied:
So, technically `GL_EXT_scalar_block_layout` is not a SPIR-V extension, correct? I tried a couple more `GL_EXT_*` extensions (e.g., `GLSL_EXT_ray_tracing`) and it looks like, for glslang at least, the `GL_*` extensions only appear in [OpSourceExtension](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpSourceExtension) instructions (which "...has no semantic impact and can safely be removed from a module"), and only `SPV_*` extensions appear in `OpExtension` instructions.

@sfricke-samsung seems like we should implement your suggestion for checking the extension prefix, and probably emit a warning (error?) message if there is a non-SPV extension found in an `OpExtension` instruction (@jeremy-lunarg?)?

sfricke-samsung replied:
ok, looked more into this and I "think" `OpExtension` with a non-SPIRV extension is invalid SPIR-V, but I will bring up with working group to decide and maybe add to `spirv-val` in the future... The working group call in Wednesday morning, so will know soon

sfricke-samsung replied:
So its internal SPIR-V issue 634

I think for the short time we are better off just doing the `SPV` string check in the validation layers

As for the warning, neither for/against personally

mark-lunarg replied:
Not up on all the details, but it seems like an error saying that it's invalid spirv is appropriate for now (maybe with a reason _why_ it's invalid?).

ncesario-lunarg commented:
@mark-lunarg just noticed you specifically called out _error_; I added a check for the extension name with a _warning_. I can change this to an error if you think that's what it should be.

mark-lunarg commented:
@ncesario-lunarg, yes please, should be an error or nothing -- the place for warnings is in Best Practices.

sfricke-samsung commented on layers/generated/spirv_validation_helper.cpp:
```diff
@@ -643,12 +643,19 @@ bool CoreChecks::ValidateShaderCapabilitiesAndExtensions(SHADER_MODULE_STATE con
+                }
+            } else {
+                skip |= LogError(device, kVUID_Core_Shader_InvalidExtension,
+                    "vkCreateShaderModule(): The SPIR-V code uses the '%s' extension which is not a SPIR-V extension", extension_name.c_str());
```
I would add a note about looking into using `OpSourceExtension` as possible correct solution

ncesario-lunarg replied:
Does this sound ok?

`vkCreateShaderModule(): The SPIR-V code uses the '%s' extension which is not a SPIR-V extension. Please use a SPIR-V extension (https://github.com/KhronosGroup/SPIRV-Registry) for OpExtension instructions. Non-SPIR-V extensions can be recorded in SPIR-V using the OpSourceExtension instruction.`

sfricke-samsung replied:
🚢

jeremy-lunarg commented on layers/shader_validation.cpp:
```diff
@@ -1291,9 +1291,17 @@ void DefineStructMember(const SHADER_MODULE_STATE &src, const spirv_inst_iter &i
+                // that "points to something."
+                data1.size = 8;
+            } else {
+                // If it's OpTypePointer. it means the member is a buffer, the type will be TypePointer, and then struct
```
"and then struct"

Are you referring to the pointee type?

ncesario-lunarg replied:
I believe so. I copied this comment from before. However, I don't understand _why_ this is true based on [the spec](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#OpTypePointer).

jeremy-lunarg replied:
As far as I understand, the only restriction on pointee-type is with respect to structure members.
"If an operand is not yet defined, it must be defined by an OpTypePointer, where the type pointed to is an
OpTypeStruct."
That said, I'm not entirely sure what that comment is implying.

ncesario-lunarg replied:
It's not clear to me if this is correct or not either. I don't see why a pointer can't point to a "non-struct type" (based solely on what's written in the spec). I'm probably misunderstanding something here, though.

ncesario-lunarg commented:
After discussing with @jeremy-lunarg, I removed the "custom compiler test" as it's not clear as to whether the spir-v is valid (based on the `OpExtension "GL_EXT_scalar_block_layout"` instruction). I also added a test for the "SPV_ extension validation" (since the test I removed was the only thing exercising it to my knowledge).

ncesario-lunarg commented:
@mark-lunarg @sfricke-samsung If this also looks good to you guys, I'll go ahead and merge this.

sfricke-samsung commented:
I am personally good with the changes

mark-lunarg commented:
🚢
